File: FREQNESS_Toolbox/FREQNESS_Functions/FREQNESS_BackProjection.m
function backProj = FREQNESS_BackProjection(FREQ, freq2project, varargin)

% ========================================================================
%
%  FREQUENCY-RESOLVED NETWORK ESTIMATION TOOLBOX
%
%  If you find this function useful, please cite the first FREQNESS paper:
%  Rosso, M., Fernández‐Rubio, G., Keller, P. E., Brattico, E., Vuust, P.,
%  Kringelbach, M. L., & Bonetti, L. (2025).
%  FREQ‐NESS Reveals the Dynamic Reconfiguration of Frequency‐Resolved Brain
%  Networks During Auditory Stimulation.
%  Advanced Science, 2413195.
%  https://doi.org/10.1002/advs.202413195
%
% ========================================================================
%  This function backprojects selected FREQNESS components from component
%  space into voxel space. The output represents the broadband voxel-space
%  contribution of networks estimated at one selected frequency.
%
%  FREQNESS_NetworkEstimation computes the component time series as
%
%      Y = W' * X
%
%  where W contains the GED eigenvectors and X is the broadband voxel
%  data. Here, a reconstruction-consistent forward model is obtained from
%  the complete set of retained eigenvectors:
%
%      A = pinv(W')
%
%  Selected network contributions are then reconstructed as:
%
%      X_network = A(:,components) * Y(components,:)
%
%  The forward model is computed before selecting components. This
%  preserves the dual relationship between each retained spatial filter
%  and its corresponding reconstruction vector.
%
%  NOTE: FREQ.pats contains target-frequency activation patterns intended
%  for spatial interpretation. These patterns are not used here because
%  their scaling is not the inverse mapping paired with FREQ.ts.
%
% ------------------------------------------------------------------------
%  INPUT ARGUMENTS:
% ------------------------------------------------------------------------
%
%  - FREQ : structure with fields
%           • FREQ.evecs -> GED eigenvectors
%                           [nVoxels x nComp x nFrex x (nSubs)]
%           • FREQ.ts    -> broadband component time series
%                           [nComp x nTime x nFrex x (nSubs)]
%           • FREQ.frex  -> frequency vector [nFrex x 1]
%
%  - freq2project : scalar frequency in Hz identifying the GED solution
%                   to backproject. If the requested value is not present
%                   in FREQ.frex, the closest available frequency is used.
%
% ------------------------------------------------------------------------
%  OPTIONAL NAME–VALUE PAIRS:
% ------------------------------------------------------------------------
%
%  - 'comps2project' : component index or vector of component indices to
%                      backproject. Default: 1. When multiple components
%                      are selected, their voxel-space contributions are
%                      summed.
%
% ------------------------------------------------------------------------
%  OUTPUT:
% ------------------------------------------------------------------------
%
%  - backProj : reconstructed broadband network activity in voxel space.
%               Its dimensions are [nVoxels x nTime] for one participant
%               and [nVoxels x nTime x nSubs] for multiple participants.
%
% ------------------------------------------------------------------------
%  AUTHORS:
%  Mattia Rosso & Leonardo Bonetti
%  mattia.rosso@clin.au.dk
%  leonardo.bonetti@clin.au.dk; leonardo.bonetti@psych.ox.ac.uk
%  Center for Music in the Brain, Aarhus University
%  Centre for Eudaimonia and Human Flourishing, Linacre College, University of Oxford
%  Aarhus (DK), Oxford (UK)
%
% ========================================================================

%% Check mandatory inputs

if nargin < 2 || isempty(freq2project)
    error(['freq2project is required. Specify one frequency in Hz from ' ...
        'the range covered by FREQ.frex.']);
end

if ~isstruct(FREQ)
    error('Input must be a FREQ structure.');
end

required_fields = {'evecs','ts','frex'};
for fieldi = 1:numel(required_fields)
    if ~isfield(FREQ,required_fields{fieldi}) || isempty(FREQ.(required_fields{fieldi}))
        error('FREQ.%s is missing or empty.',required_fields{fieldi});
    end
end

if ~isnumeric(freq2project) || ~isscalar(freq2project) || ~isfinite(freq2project)
    error('freq2project must be one finite scalar frequency expressed in Hz.');
end

%% Handle optional arguments (name-value pairs)

opts = struct('comps2project', []);
opts = parse_name_value_pairs(opts, varargin{:});

comps2project = opts.comps2project;

if isempty(comps2project)
    comps2project = 1;
    fprintf('\nFREQNESS BackProjection: no components specified. Defaulting to component #1.\n');
end

%% Map and validate inputs from FREQ

frex  = FREQ.frex(:);
evecs = FREQ.evecs;
ts    = FREQ.ts;

nvoxs  = size(evecs,1);
ncomps = size(evecs,2);
nfrex  = size(evecs,3);
nsubs  = size(evecs,4);

if numel(frex) ~= nfrex
    error('Length of FREQ.frex does not match the 3rd dimension of FREQ.evecs.');
end

if size(ts,1) ~= ncomps || size(ts,3) ~= nfrex || size(ts,4) ~= nsubs
    error('FREQ.evecs and FREQ.ts have incompatible dimensions.');
end

if ~isnumeric(comps2project) || ~isvector(comps2project) || ...
        any(~isfinite(comps2project)) || any(comps2project ~= round(comps2project))
    error('comps2project must contain finite integer component indices.');
end

comps2project = unique(comps2project(:)','stable');

if any(comps2project < 1) || any(comps2project > ncomps)
    error('comps2project must contain indices between 1 and %d.',ncomps);
end

% Map requested frequency to the closest frequency analyzed by FREQNESS
[~, which_frex] = min(abs(frex - freq2project));
freq2project_actual = frex(which_frex);

tol = 1e-10;
if ~ismembertol(freq2project,frex,tol)
    warning(['Requested freq2project (%.3f Hz) is not present in FREQ.frex. ' ...
        'Using the closest available frequency: %.3f Hz.'], ...
        freq2project,freq2project_actual);
end

fprintf(['\nFREQNESS BackProjection: backprojecting component(s) %s at ' ...
    '%.3f Hz for %d participants.\n'], ...
    mat2str(comps2project),freq2project_actual,nsubs);

%% Backproject selected components into voxel space

ntime = size(ts,2);
backProj = zeros(nvoxs,ntime,nsubs);

for subi = 1:nsubs

    % Complete retained GED filter set for this frequency and participant
    W = reshape(evecs(:,:,which_frex,subi),nvoxs,ncomps);

    % Reconstruction-consistent forward model paired with Y = W' * X
    A = pinv(W');

    % Stored broadband component time series
    Y = reshape(ts(:,:,which_frex,subi),ncomps,ntime);

    % Summed voxel-space contribution of the selected networks
    backProj(:,:,subi) = A(:,comps2project) * Y(comps2project,:);

end

% Preserve a 2D output for single-participant inputs
if nsubs == 1
    backProj = backProj(:,:,1);
end

end


%% Helper Function: Parse Name-Value Pairs
function opts = parse_name_value_pairs(opts, varargin)
if mod(length(varargin), 2) ~= 0
    error('Arguments must be given as name-value pairs.');
end
for i = 1:2:length(varargin)
    name = lower(varargin{i});
    if isfield(opts, name)
        opts.(name) = varargin{i+1};
    else
        error(['Unrecognized argument: ', name]);
    end
end
end

%%

File: FREQNESS_Toolbox/python/src/freqness/FREQNESS_BackProjection.py
"""Backproject selected FREQ-NESS components into voxel space.

If you use this toolbox, please cite:
Rosso, M., Fernández-Rubio, G., Keller, P. E., Brattico, E., Vuust, P.,
Kringelbach, M. L., & Bonetti, L. (2025). FREQ-NESS Reveals the Dynamic
Reconfiguration of Frequency-Resolved Brain Networks During Auditory
Stimulation. Advanced Science, 2413195.
https://doi.org/10.1002/advs.202413195

Toolbox authors
---------------
Mattia Rosso - Center for Music in the Brain, Aarhus University -
mattia.rosso@clin.au.dk
Chiara Malvaso - University of Bologna - chiara.malvaso2@unibo.it
Leonardo Bonetti - Center for Music in the Brain, Aarhus University; Centre
for Eudaimonia and Human Flourishing, Linacre College, University of Oxford -
leonardo.bonetti@clin.au.dk; leonardo.bonetti@psych.ox.ac.uk
"""

from __future__ import annotations

from typing import Any, Mapping
import warnings

import numpy as np
from numpy.typing import ArrayLike, NDArray
from scipy.linalg import pinv


FloatArray = NDArray[np.float64]


def _field(container: Any, name: str, *, required: bool = True) -> Any:
    if isinstance(container, Mapping):
        if name in container:
            return container[name]
    elif hasattr(container, name):
        return getattr(container, name)
    if required:
        raise ValueError(f"FREQ.{name} is missing or empty")
    return None


def _real_finite_array(value: Any, name: str) -> FloatArray:
    array = np.asarray(value)
    if array.size == 0:
        raise ValueError(f"{name} is missing or empty")
    if not np.issubdtype(array.dtype, np.number) or np.iscomplexobj(array):
        raise TypeError(f"{name} must contain real numeric values")
    array = np.asarray(array, dtype=float)
    if not np.all(np.isfinite(array)):
        raise ValueError(f"{name} must contain finite values")
    return array


def _frequency_value(value_to_check: Any, name: str = "freq2project") -> float:
    value = np.asarray(value_to_check)
    if (
        value.ndim != 0
        or not np.issubdtype(value.dtype, np.number)
        or np.iscomplexobj(value)
        or isinstance(value_to_check, (bool, np.bool_))
    ):
        raise TypeError(f"{name} must be one positive finite scalar frequency")
    frequency = float(value)
    if not np.isfinite(frequency) or frequency <= 0:
        raise ValueError(f"{name} must be one positive finite scalar frequency")
    return frequency


def _component_indices(
    comps2project: ArrayLike | None,
    ncomponents: int,
    name: str = "comps2project",
) -> tuple[NDArray[np.int64], NDArray[np.int64]]:
    if comps2project is None:
        one_based = np.array([1], dtype=np.int64)
    else:
        raw = np.asarray(comps2project)
        if raw.size == 0:
            one_based = np.array([1], dtype=np.int64)
        else:
            if raw.ndim == 0:
                raw = raw.reshape(1)
            if (
                raw.ndim != 1
                or not np.issubdtype(raw.dtype, np.number)
                or np.iscomplexobj(raw)
                or np.issubdtype(raw.dtype, np.bool_)
                or not np.all(np.isfinite(raw))
                or not np.all(raw == np.round(raw))
            ):
                raise ValueError(
                    f"{name} must contain finite integer component indices"
                )

            # MATLAB uses unique(..., 'stable'). Preserve the public order while
            # retaining MATLAB-compatible one-based component numbering.
            stable_unique = list(dict.fromkeys(raw.astype(np.int64).tolist()))
            one_based = np.asarray(stable_unique, dtype=np.int64)

    if np.any(one_based < 1) or np.any(one_based > ncomponents):
        raise ValueError(
            f"{name} must contain indices between 1 and {ncomponents}"
        )
    return one_based, one_based - 1


def _component_text(component_numbers: NDArray[np.int64]) -> str:
    if component_numbers.size == 1:
        return str(int(component_numbers[0]))
    return "[" + " ".join(str(int(value)) for value in component_numbers) + "]"


def _canonical_inputs(
    FREQ: Any,
) -> tuple[FloatArray, FloatArray, FloatArray, FloatArray]:
    frequencies = _real_finite_array(_field(FREQ, "frex"), "FREQ.frex")
    if frequencies.ndim != 1:
        raise ValueError("FREQ.frex must be a one-dimensional vector")
    if np.any(frequencies <= 0):
        raise ValueError("FREQ.frex must contain positive frequencies")

    eigenvectors = _real_finite_array(_field(FREQ, "evecs"), "FREQ.evecs")
    time_series = _real_finite_array(_field(FREQ, "ts"), "FREQ.ts")

    if eigenvectors.ndim == 3:
        eigenvectors = eigenvectors[:, :, :, np.newaxis]
    if time_series.ndim == 3:
        time_series = time_series[:, :, :, np.newaxis]
    if eigenvectors.ndim != 4:
        raise ValueError(
            "FREQ.evecs must have shape "
            "(voxels, components, frequencies, participants)"
        )
    if time_series.ndim != 4:
        raise ValueError(
            "FREQ.ts must have shape "
            "(components, time, frequencies, participants)"
        )

    nvoxels, ncomponents, nfrequencies, nsubjects = eigenvectors.shape
    if min(nvoxels, ncomponents, nfrequencies, nsubjects) == 0:
        raise ValueError("FREQ.evecs dimensions must be non-empty")
    if frequencies.size != nfrequencies:
        raise ValueError(
            "Length of FREQ.frex does not match the frequency dimension "
            "of FREQ.evecs"
        )
    if (
        time_series.shape[0] != ncomponents
        or time_series.shape[1] == 0
        or time_series.shape[2] != nfrequencies
        or time_series.shape[3] != nsubjects
    ):
        raise ValueError("FREQ.evecs and FREQ.ts have incompatible dimensions")

    raw_scales = _field(FREQ, "scale_factors", required=False)
    if raw_scales is None:
        scale_factors = np.ones(nsubjects, dtype=float)
    else:
        scale_factors = _real_finite_array(
            raw_scales,
            "FREQ.scale_factors",
        ).reshape(-1)
        if scale_factors.size != nsubjects:
            raise ValueError(
                "FREQ.scale_factors must contain one value per participant"
            )
        if np.any(scale_factors <= 0):
            raise ValueError("FREQ.scale_factors must contain positive values")

    return frequencies, eigenvectors, time_series, scale_factors


def FREQNESS_BackProjection(
    FREQ: Any,
    freq2project: float,
    *,
    comps2project: ArrayLike | None = None,
) -> FloatArray:
    """Reconstruct selected FREQ-NESS networks in voxel space.

    The reconstruction uses the dual forward model paired with the stored
    broadband component time series::

        Y = W.T @ X
        A = pinv(W.T)
        X_network = A[:, components] @ Y[components, :]

    ``comps2project`` deliberately uses MATLAB-compatible one-based component
    numbering. When the requested frequency is absent, the closest analyzed
    frequency is used. If network estimation rescaled a participant internally,
    its stored ``scale_factors`` value is removed so the result is returned in
    the units of the original input data.

    A single participant produces ``(voxels, time)`` output; multiple
    participants produce ``(voxels, time, participants)`` output. When fewer
    components than voxels were retained, selecting all retained components
    reconstructs the data's projection onto that retained GED subspace.
    """
    requested_frequency = _frequency_value(freq2project)
    if comps2project is None or np.asarray(comps2project).size == 0:
        print(
            "\nFREQNESS BackProjection: no components specified. Defaulting to component #1."
        )
    frequencies, eigenvectors, time_series, scale_factors = _canonical_inputs(
        FREQ
    )
    component_numbers, component_indices = _component_indices(
        comps2project,
        eigenvectors.shape[1],
    )

    frequency_index = int(np.argmin(np.abs(frequencies - requested_frequency)))
    actual_frequency = float(frequencies[frequency_index])
    if not np.isclose(
        requested_frequency,
        actual_frequency,
        atol=1e-10,
        rtol=0.0,
    ):
        warnings.warn(
            f"Requested freq2project ({requested_frequency:.3f} Hz) is not "
            f"present in FREQ.frex; using the closest available frequency: "
            f"{actual_frequency:.3f} Hz.",
            UserWarning,
            stacklevel=2,
        )

    nvoxels = eigenvectors.shape[0]
    ntime = time_series.shape[1]
    nsubjects = eigenvectors.shape[3]
    print(
        "\nFREQNESS BackProjection: backprojecting component(s) "
        f"{_component_text(component_numbers)} at {actual_frequency:.3f} Hz "
        f"for {nsubjects} participants."
    )
    back_projection = np.empty((nvoxels, ntime, nsubjects), dtype=float)

    for subject in range(nsubjects):
        filters = eigenvectors[:, :, frequency_index, subject]
        forward_model = pinv(filters.T, check_finite=False)
        components = time_series[:, :, frequency_index, subject]
        reconstruction = (
            forward_model[:, component_indices]
            @ components[component_indices, :]
        )
        back_projection[:, :, subject] = reconstruction / scale_factors[subject]

    if nsubjects == 1:
        return back_projection[:, :, 0]
    return back_projection


__all__ = ["FREQNESS_BackProjection"]
